Feat: OAuth 2.0 Authentication plugin (#1215)
Kong/developer.konghq.com: app/_kong_plugins/oauth2/index.md

     --data "response_type=XXX" \
     --data "scope=XXX" \
     --data "provision_key=XXX" \
     --data "authenticated_userid=XXX"
   ```

   The `provision_key` is the key the plugin generates when it's added to the Service. `authenticated_userid` is the ID of the logged-in end user who grants the permission.

1. {{site.base_gateway}} responds with a JSON response:

   ```json
   {
     "redirect_uri": "http://some/url"
   }
   ```

   The response can be either a `200 OK` or `400 Bad Request`.
1. In both cases, ignore the response status code and just redirect the user to whatever URI is being returned in the `redirect_uri` property.

1. The client application takes it from here, and continues the flow with {{site.base_gateway}} with no other interaction with your web application. Like exchanging the authorization code for an access token if it's an Authorization Code Grant flow.

1. Once the Access Token is retrieved, the client application makes requests on behalf of the user to your upstream service.

1. Access Tokens can expire, and when that happens the client application needs to renew the Access Token with {{site.base_gateway}} and retrieve a new one.

> In this flow, the steps that you need to implement are:
> * The login page (step 2)
> * The Authorization page, with its backend that simply collects the values, makes a `POST` request to {{site.base_gateway}} and redirects the user to whatever URL {{site.base_gateway}} has returned (steps 3 to 7)

Reviewer: ```suggestion
In this flow, the steps that you need to implement are:
* The login page (step 2)
* The Authorization page, with its backend that simply collects the values, makes a `POST` request to {{site.base_gateway}}, and redirects the user to whatever URL {{site.base_gateway}} has returned (steps 3 to 7)
```

## gRPC requests

The same access tokens can be used by gRPC applications:
```bash
grpcurl -H 'authorization: bearer XXX' ...
```

> Note that the rest of the credentials flow uses HTTPS and not gRPC. Depending on your application, you may have to configure the `oauth2` plugin on two separate Routes: one under `protocols: ["https"]` and another under `protocols: ["grpcs"]`.

Reviewer: ```suggestion
The rest of the credentials flow uses HTTPS and not gRPC. Depending on your application, you may have to configure the `oauth2` plugin on two separate Routes: one under `protocols: ["https"]` and another under `protocols: ["grpcs"]`.
```

## WebSocket requests

This plugin can't issue new tokens from a WebSocket Route, because the request
will be rejected as an invalid WebSocket handshake. Therefore, to use this
plugin with WebSocket services, an additional non-WebSocket Route must be used
to issue tokens. For more details, see [Enable OAuth 2.0 authentication for WebSocket requests with {{site.base_gateway}}](/how-to/enable-oauth2-authentication-for-websocket-requests/)

## Resource owner password credentials (legacy systems only)

{:.warning}
 > **Important**: The OAuth2 Security Best Practice explicitly mentions that Resource Owner Password Credentials
  ["MUST NOT BE USED"](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-13#section-3.4).
  The following section is left here as a reference for supporting legacy systems.

> The [Resource Owner Password Credentials Grant](https://tools.ietf.org/html/rfc6749#section-4.3) is a much simpler version of the Authorization Code flow, but it still requires to build an authorization backend (without the frontend) in order to make it work properly.

Reviewer: ```suggestion
The [Resource Owner Password Credentials Grant](https://tools.ietf.org/html/rfc6749#section-4.3) is a much simpler version of the Authorization Code flow, but it still requires to build an authorization backend (without the frontend) to make it work properly.
```

1. On the first request, the client application makes a request with some OAuth2 parameters to your web application.

2. The backend of your web application adds the `provision_key` and `grant_type` parameters to the parameters originally sent by the client, then makes a `POST` request to {{site.base_gateway}} using the `/oauth2/token` endpoint of the configured plugin. If an `Authorization` header is sent by the client, that must be added too. For example:

    ```bash
    curl https://your.service.com/oauth2/token \
      --header "Authorization: Basic czZCaGRSa3F0MzpnWDFmQmF0M2JW" \
      --data "client_id=XXX" \
      --data "client_secret=XXX" \
      --data "grant_type=password" \
      --data "scope=XXX" \
      --data "provision_key=XXX" \
      --data "authenticated_userid=XXX" \
      --data "username=XXX" \
      --data "password=XXX"
    ```

    The `provision_key` is the key the plugin has generated when it has been added to the Service, while `authenticated_userid` is the ID of the end user whose `username` and `password` belong to.

> 3. {{site.base_gateway}} will respond with a JSON response

Reviewer: ```suggestion
3. {{site.base_gateway}} responds with a JSON response.
```
